prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs

```diff
@@ -1733,16 +1728,16 @@ layoutSmart = layoutWadlerLeijen
     -- exponential runtime (and is prohibitively expensive in practice).
     fits :: PageWidth
          -> Int -- ^ Minimum nesting level to fit in
-         -> Int -- ^ Width in which to fit the first line
+         -> Maybe Int -- ^ Width in which to fit the first line
          -> SimpleDocStream ann
          -> Bool
-    fits _ _ w _ | w < 0                    = False
+    fits _ _ (Just w) _ | w < 0             = False
     fits _ _ _ SFail                        = False
     fits _ _ _ SEmpty                       = True
-    fits pw m w (SChar _ x)                 = fits pw m (w - 1) x
-    fits pw m w (SText l _t x)              = fits pw m (w - l) x
+    fits pw m w (SChar _ x)                 = fits pw m (pred <$> w) x
+    fits pw m w (SText l _t x)              = fits pw m (subtract l <$> w) x
     fits pw m _ (SLine i x)
-      | m < i, AvailablePerLine cpl _ <- pw = fits pw m (cpl - i) x
+      | m < i, AvailablePerLine cpl _ <- pw = fits pw m (Just (cpl - i)) x
       | otherwise                           = True
```

Comment:
What I don't understand is why it doesn't seem to be necessary to continue checking for the absence of `SFail` when using the `Unbounded` layout – at least the property tests don't seem to mind.